docs/design-proposals/node/Unified-Vertical-Scaling-for-Container-and-VM-Pods.md

```diff
@@ -0,0 +1,646 @@
+  Pods in the Namespace don't exceed quota,
+* PodResourceAllocation admission plugin is ordered before ResourceQuota.
+* for LimitRanger we check that a resize request does not violate the min and
+  max limits specified in LimitRange for the Pod's namespace.
+
+Kubelet:
+* set Pod's Status.ContainerStatuses[i].Resources for Containers upon placing
+  a new Pod on the Node,
+* update Pod's Spec.Containers[i].ResourcesAllocated upon resize,
+* change UpdateContainerResources CRI API to work for both Linux & Windows.
+
+Scheduler:
+* compute resource allocations using Pod.Spec.Containers[i].ResourcesAllocated.
+
+Other components:
+* check how the change of meaning of resource requests influence other
+  Kubernetes components.
+
+### Changes in Arktos-vm-runtime-service
+
+#### Background and context
+Currently in Arktos, the VM workload POD is defined the same way as container pod in terms of resources in the spec, i.e.
+
+    spec:
+      resource:
+       limits:
+         cpu:
+         memory:
+       request:
+         cpu:
+         memory:
+     
+Arktos controls the resource the same way as for containers, i.e. schedule based on requested resource level, and evict 
+POD based on the limits values, per some defined eviction policies.
+
+Unlike the container, which is essentially the isolated process on the host, there are quite some differences for VM that
+are worth to describe here to lay some common ground before the design for supporting vertical scaling for VMs.
+
+For VM types, the resources for VM are also controlled by the Hypervisor and the management layer i.e. the Libvirt and the 
+Qemu/KVM/Xen. For memory resource allocation, a few essential configurations are in the Libvirt domain definition:
+
+    <domain>
+      ...
+      <maxMemory slots='16' unit='KiB'>1524288</maxMemory>
+      <memory unit='KiB'>524288</memory>
+      <currentMemory unit='KiB'>524288</currentMemory>
+      ...
+    </domain>
+
+Memory defines the max memory allocation at the boot time, while the currentMemory defines the real memory allocation for
+the guest OS. The delta between the currentMemory and memory is the space for memory ballooning to "reclaim" memory
+between VMs and the hypervisor. The maxMemory defines the max memory allocation for a VM at runtime, which can be achieved
+with memory hotplug.
+
+Please be noted that there are a few limitations in current design and workflow:
+1. Only spec.resource.limit was passed to the runtime from Kubelet via the linuxContainerResources.MemoryLimitInBytes
+2. The linuxContainerResources.MemoryLimitInBytes was used to set the "memory" in the VM domain xml definition. currentMemory
+   is set to the "memory" setting by default.
+3. Ballooning operation, i.e. to adjust the currentMemory, an explict API call "setmem" need to be called, i.e. no auto 
+   adjust for currentMemory based on the application running in the VM
+4. Adjust "memory" can only be done after VM is restarted
```

Comment:
I'm confused by item #4.  Is it about hotplug or balloning? But neither of them require restart?

Reply:
it is about ballooning. with hotplug new memory device, it won't require a restart. and the libvirt/hypervisor honors the new memory devices to the VM.  Ballooning cannot scale-up to the size which "memory" config sets to.